blockstore/splitstore/splitstore.go

```diff
@@ -936,6 +936,12 @@ func (s *SplitStore) doWarmup(curTs *types.TipSet) error {
 func (s *SplitStore) compact(curTs *types.TipSet, wg *sync.WaitGroup) {
 	log.Info("waiting for active views to complete")
 	start := time.Now()
+	// I'm not holding the txnLk here. Is it not possible for:
+	//
+	// 1. The waitgroup to temporarily hit 0.
+	// 2. protectView to be called, raising it back to 1.
+	//
+	// _Reusing_ a waitgroup is legal. Hitting 0 then going back to 1 while "waiting" is not.
```

Comment:
I.e., https://pkg.go.dev/sync#WaitGroup.Add

One solution here is to keep a separate count (under a lock)? Or maybe using a waitgroup was a bad idea and we should just use mutex, condition, and count.

"Wait"

```go
mu.Lock()
waitingToCompact = true
for activeViews > 0 {
    cond.Wait()
}
waitingToCompact = false
// do any other state updates we need to do atomically (e.g., begin the transaction?)
mu.Unlock()
```

"View" enter:

```go
mu.Lock()
defer mu.Unlock()
if waitingToCompact && activeViews == 0 {
    // abort
}
activeViews++
```

"View" exit

```go
mu.Lock()
defer mu.Unlock()
activeViews--
if activeViews == 0 && waitingToCompact {
    cond.Broadcast()
}
```

Reply:
But maybe there are easier solutions.

Reply:
and finetuned in https://github.com/filecoin-project/lotus/pull/6474/commits/af399529ecca4c6e082474cada76c9b96a3d8f86